### PR TITLE
Adding docs on the various extension mechanisms available in IREE.

### PR DESCRIPTION
**Pull request by benvanik**
I tried to keep this relatively high-level for now (mostly because
many of these mechanisms are not fully implemented) but it's such
a common question that we really need someplace to direct people to
for the 10,000ft view and capabilities we (will) provide.

There's just the single design overview for now but as we build and
document these mechanisms we can add user-facing docs for the ones
that people will encounter with samples/etc.

### CODE REVIEW

File: docs/website/docs/extensions/index.md [hunk 1]
```diff
@@ -0,0 +1,442 @@
+# Extension mechanisms
+
+**NOTE**: much of this describes provisions for extension within IREE but until
+the core of the system has settled little work will be done to fully flesh-out
+and document them in detail. A large majority of things that would make someone
+want to extend IREE can instead be accomplished much easier and performantly
+using native MLIR dialects that are then processed by the IREE compiler.
```
Comment (ScottTodd) [inline docs/website/docs/extensions/index.md:7]: if you want, you can use an admonition (https://squidfunk.github.io/mkdocs-material/reference/admonitions/) for this

image


```suggestion
!!! Note
    Much of this describes provisions for extension within IREE but until
    the core of the system has settled little work will be done to fully
    flesh-out and document them in detail. A large majority of things that would
    make someone want to extend IREE can instead be accomplished much easier and
    performantly using native MLIR dialects that are then processed by the IREE
    compiler.
```

File: docs/website/docs/extensions/index.md [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+IREE has a compiler and runtime separation, a multi-layered architecture, and
+split between execution of "host code" that schedules compute-heavy work and
+[SPMD](https://en.wikipedia.org/wiki/SPMD) "device code" that performs the bulk
+of compute operations. Each axis has a different set of extension mechanisms
+that can be used independently or combined.
+
+The primary goal when extending any framework should first be to avoid extending
+it at all. There is no mechanism that is free - whether in terms of engineering
```
Comment (ScottTodd) [inline docs/website/docs/extensions/index.md:16]: This appears as a pretty big wall of text on the rendered site:

image

Diagrams, images, bullet points, emojis/icons, admonitions (like "tip" or "warning") or more paragraph splits could help break it up and make it easier to scan. I haven't done a full read through yet, so I may come up with more concrete suggestions later.

**Review by ScottTodd — COMMENTED** [review]
Just a few high level comments on website style after a quick look. Will take a more detailed look through tomorrow.

File: docs/website/docs/extensions/index.md [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+  interfaces.
+
+### When to use
+
+* :heavy_check_mark: Code that must be emitted during target lowering - such as
+  something optimizing for a particular CPU architecture.
+* :heavy_check_mark: Hot code mixed with generated code at a fine granularity
+  (within the innermost loop).
+* :x: External existing hand-authored libraries. Either statically or
+  dynamically link instead.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply docs/website/docs/extensions/index.md:266]: material-check looks nice - is there an x you want instead that matches it?

[the remark this answers is not in the document]
Comment (ScottTodd) [reply docs/website/docs/extensions/index.md:266]: how about `:material-close:` or `:material-cancel:`?

image

File: docs/website/docs/extensions/index.md [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+* No IREE compiler code changes required.
+* Produced artifacts are portable across IREE deployment configurations.
+* Full system access is allowed - the VM just calls external functions.
+* Runtime modules can be implemented (via shims) in other languages/runtimes.
+
+### Cons
+
+* _(Today)_ no asynchronous scheduling or pipelining.
+  * Future improvements will enable pipelining of host calls.
```
Comment (ScottTodd) [inline docs/website/docs/extensions/index.md:173]: nit: mkdocs nested lists should have four spaces of indentation between levels
```suggestion
* _(Today)_ no asynchronous scheduling or pipelining.
    * Future improvements will enable pipelining of host calls.
```

(throughout this page)

File: docs/website/docs/extensions/index.md [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+
+### Implementation
+
+The compiler is changed to produce calls to imports via a dynamic import table
+provided to each dispatch function. The import table is declared in the
+executable library for use at runtime. Runtime applications register an import
+provider to resolve named symbols in the import table to C functions that
+marshal arguments and results.
+
+The compiler-side needs some additional work but an example is included here: https://github.com/google/iree/issues/7504.
```
Comment (ScottTodd) [inline docs/website/docs/extensions/index.md:440]: nit: mkdocs does not make a hyperlink out of loose URLs :( you'll need to wrap in `[description](url)`

File: docs/website/docs/extensions/index.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+minimal overhead and is used for example [by the IREE HAL itself](https://github.com/google/iree/tree/main/iree/modules/hal).
+There is also a C++ wrapper that is significantly easier to work with however it
+needs some performance improvements.
+
+A full end-to-end example can be found under [`iree/samples/custom_modules/`](https://github.com/google/iree/tree/main/iree/samples/custom_modules),
+though it should not currently be considered representative of best practices.
+
+## 3. Extend target-specific device conversion patterns
+
+* TL;DR: add patterns to `iree/Compiler/Codegen/` to emit target code.
```
Comment (ScottTodd) [inline docs/website/docs/extensions/index.md:210]: just an idea (which I'm not sold on): you could use an admonition for these TLDRs to get a bit of color/styling

image
```suggestion
## 3. Extend target-specific device conversion patterns

!!! tip
    TL;DR: add patterns to `iree/Compiler/Codegen/` to emit target code.
```

you could also do that for the quote:

image

File: docs/website/docs/extensions/index.md (hunk 2 above)
Comment (ScottTodd) [reply docs/website/docs/extensions/index.md:16]: Some more thoughts after reading through:

The words here are all important for providing context and high level guidance on pros/cons of various approaches. If someone is seriously considering an extension, they should take the time to read through and evaluate each option. That being said, breaking up into bullet points or sections ("why extend?", "performance fundamentals", etc.) would still help with scanning.

I'm fine with landing as-is though.

**Review by ScottTodd — APPROVED** [review]
LGTM after a few style tweaks.

A few more links to implementation examples like the custom modules sample would help I think. Or maybe concrete walkthroughs like "my model has an FFT op, I see an FFT library, what should I do?".

File: docs/website/docs/extensions/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (benvanik) [reply docs/website/docs/extensions/index.md:266]: material-close it seems

File: docs/website/docs/extensions/index.md (hunk 2 above)
Comment (benvanik) [reply docs/website/docs/extensions/index.md:16]: will split up into sections like you recommend below
